src/ddoc_cache/test/ddoc_cache_open_test.erl

```diff
@@ -0,0 +1,98 @@
+% Licensed under the Apache License, Version 2.0 (the "License"); you may not
+% use this file except in compliance with the License. You may obtain a copy of
+% the License at
+%
+%   http://www.apache.org/licenses/LICENSE-2.0
+%
+% Unless required by applicable law or agreed to in writing, software
+% distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+% WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+% License for the specific language governing permissions and limitations under
+% the License.
+
+-module(ddoc_cache_open_test).
+
+-export([
+    dbname/1,
+    ddocid/1,
+    recover/1,
+    insert/2
+]).
+
+-include_lib("couch/include/couch_db.hrl").
+-include_lib("eunit/include/eunit.hrl").
+-include("ddoc_cache_test.hrl").
+
+
+%% behaviour callbacks
+dbname(DbName) ->
+    DbName.
+
+ddocid(_) ->
+    no_ddocid.
+
+recover({deleted, _DbName}) ->
+    erlang:error(database_does_not_exist);
+recover(DbName) ->
+    ddoc_cache_entry_validation_funs:recover(DbName).
+
+insert(_, _) ->
+    ok.
+
+start_couch() ->
+    Ctx = ddoc_cache_tutil:start_couch(),
+    meck:new(ddoc_cache_entry_validation_funs, [passthrough]),
+    meck:expect(ddoc_cache_entry_validation_funs, recover,
+        ['_'], meck:passthrough()),
+    Ctx.
+
+
+stop_couch(Ctx) ->
+    meck:unload(),
+    ddoc_cache_tutil:stop_couch(Ctx).
+
+
+check_open_error_test_() ->
+    {
+        setup,
+        fun start_couch/0,
+        fun stop_couch/1,
+        {
+            with, [
+                fun should_return_database_does_not_exist/1,
+                fun should_not_call_recover_when_database_does_not_exist/1,
+                fun should_call_recover_when_needed/1,
+                fun should_not_crash_lru_process/1
+            ]
+        }
+    }.
+
+
+should_return_database_does_not_exist({DbName, _}) ->
+    ?assertError(
+        database_does_not_exist,
+        ddoc_cache_lru:open({?MODULE, {deleted, DbName}})).
+
+should_not_call_recover_when_database_does_not_exist({DbName, _}) ->
+    meck:reset(ddoc_cache_entry_validation_funs),
+    ?assertError(
+       database_does_not_exist,
+       ddoc_cache_lru:open({?MODULE, {deleted, DbName}})),
+    ?assertError(
+        timeout,
+        meck:wait(1, ddoc_cache_entry_validation_funs, recover, '_', 500)).
```

Comment:
@iilyak I'm still seeing this test always take > 500ms:
```
$ make eunit skip_tests+=couch_epep apps=ddoc_cache suites=ddoc_cache_open_test
======================== EUnit ========================
module 'ddoc_cache_open_test'
Application crypto was left running!
  ddoc_cache_open_test: check_open_error_test_...[0.002 s] ok
  ddoc_cache_open_test: check_open_error_test_...[0.501 s] ok
  ddoc_cache_open_test: check_open_error_test_...[0.015 s] ok
  ddoc_cache_open_test: check_open_error_test_...ok
  [done in 0.530 s]
=======================================================
  All 4 tests passed.
```
Do we really need to wait this long for that timeout?

Reply:
I don't understand how it can be related to `meck`. Here is a result of my investigation:
- `meck:wait/5` is dispatched to `meck:wait/6` [see here](https://github.com/eproxus/meck/blob/master/src/meck.erl#L570)
- `meck:wait/6` call is dispatched to `meck_proc:wait/6` [see here](https://github.com/eproxus/meck/blob/master/src/meck.erl#L593)
- `meck_proc:wait/6` does a `gen_server` call [here](https://github.com/eproxus/meck/blob/master/src/meck_proc.erl#L161)
- the `handle_call` for wait message has following shape 
  ```
    case times_called(OptFunc, ArgsMatcher, OptCallerPid, History) of
        CalledSoFar when CalledSoFar >= Times ->
             {reply, ok, S};
  ```
- the `times_called/4` function is pretty simple and just [iterate through `History` term](https://github.com/eproxus/meck/blob/master/src/meck_proc.erl#L603:L612)  

The only explanation I have is it takes time to process `ddoc_cache_lru:open` and detect the call on `meck` side. 

I don't like reducing the timeout since there is a high chance of turning the test into a flaky one. Because shorter timeout could be a problem for slower CI systems.

I'll try to reduce the timeout locally to see if it has any effect on the run time of the test suite. If it does it would mean that I don't read `meck` code correctly. If it doesn't not it would mean that the test is slow.

Reply:
JFYI: It looks like there is a way.

Reply:
I ended up reducing the timeout value